structure/redblacktree/rbtree.go

```diff
@@ -0,0 +1,404 @@
+					// Case 2
+					t.rightRotate(parentNode)
+					currNode = parentNode
+				}
+				// Case 3
+				t.leftRotate(grandNode)
+				currNode.parent.isRed = false
+				grandNode.isRed = true
+			}
+		}
+	}
+}
+
+// Simplest case, make parent level black
+func case1(parentNode, uncleNode, grandNode *Node) *Node {
+	parentNode.isRed = false
+	uncleNode.isRed = false
+	grandNode.isRed = true
+	return grandNode
+}
+
+// A left rotate similar to binary search tree. Coloring is not handled
+func (t *RBTree) leftRotate(parentNode *Node) {
+	// Swap nodes and parents
+	var currNode = parentNode.right
+	var tempNode = currNode.left
+	currNode.left = parentNode
+	parentNode.right = tempNode
+	currNode.parent = parentNode.parent
+	parentNode.parent = currNode
+
+	if t.isRoot(currNode) {
+		t.Root = currNode
+		return
+	}
+	// Set new parent left or right
+	if currNode.parent.left == parentNode {
+		currNode.parent.left = currNode
+	} else {
+		currNode.parent.right = currNode
+	}
+}
+
+// Right rotate, similar to left rotate above
+func (t *RBTree) rightRotate(parentNode *Node) {
+	var currNode = parentNode.left
+	var tempNode = currNode.right
+	currNode.right = parentNode
+	parentNode.left = tempNode
+	currNode.parent = parentNode.parent
+	parentNode.parent = currNode
+
+	if currNode.parent == nil {
+		t.Root = currNode
+		return
+	}
+	if currNode.parent.left == parentNode {
+		currNode.parent.left = currNode
+	} else {
+		currNode.parent.right = currNode
+	}
+}
+
+// Simply return new node
+func newNode(Val int, parent *Node) *Node {
+	return &Node{Val: Val, left: nil, right: nil, parent: parent, isRed: true}
+}
+
+// Public function to print preorder
+func (t *RBTree) PrintPreorder() {
+	var p = preorder(t.Root)
+	fmt.Printf("Length %d\n", len(p))
+	for _, a := range p {
+		fmt.Printf("%d %v\n", a.Val, a.isRed)
+	}
+}
+
+// Public function to print inorder
+func (t *RBTree) PrintInorder() {
+	var p = inorder(t.Root)
+	fmt.Printf("Length %d\n", len(p))
+	for _, a := range p {
+		fmt.Printf("%d %v\n", a.Val, a.isRed)
+	}
+}
+
+// Returns inorder representation
+func inorder(currNode *Node) []*Node {
+	var ret = []*Node{}
+	if currNode == nil {
+		return ret
+	}
+
+	ret = append(ret, inorder(currNode.left)...)
+	ret = append(ret, currNode)
+	ret = append(ret, inorder(currNode.right)...)
+	return ret
+}
+
+// Returns preorder representation
+func preorder(currNode *Node) []*Node {
+	var ret = []*Node{}
+	if currNode == nil {
+		return ret
+	}
+	ret = append(ret, currNode)
+	ret = append(ret, preorder(currNode.left)...)
+	ret = append(ret, preorder(currNode.right)...)
+	return ret
+}
+
+// Public search function based on value
+func (t *RBTree) Search(val int) *Node {
+	return search(t.Root, val)
+}
+
+// Internal search function, recursive search
+func search(node *Node, val int) *Node {
+	if node == nil {
+		return nil
+	}
+	if node.Val == val {
+		return node
+	}
+	if val < node.Val {
+		return search(node.left, val)
+	}
+	return search(node.right, val)
+}
```

Comment:
Just put it inside the Search function. No need to create an intermediate function if there are no more than 1 search implementation. Creating an intermediate function should only be considered if there is more than one way to do one thing and you want to give the user choice of choosing which one to do. Right now, you only have one search and it is highly unlikely anyone is going to add more searching algorithms on this.

Reply:
Yeah I was on the fence about allowing Search on any Node.. might as well add it and allow subtree search

Reply:
I'd still prefer to keep both a simple tree-wide search, that only takes int and another search that takes a node. If you want I'll expose `search` as `SearchNode` or `SearchSubTree`.

Reply:
Yeah, if you have more than one implementation of search then it is feasible that you can create an intermediate one. Note that you can utilise the power of Go by using Function as an argument. Assuming you have two different search functions with the same signature (example `searchABC(node *Node, val int) *Node` and `searchDEF(node *Node, val int) *Node` then you can have your exposed Search implementation be such that it accepts the function as an argument but since they are simple and your Search function is doing nothing but calling searchABC or searchDEF then it is fine to just expose these two functions instead of a search function. Makes it clear what the intent of the function is.

Reply:
I have just made `search` public as `SearchFromNode` - basically allowing search within a subtree. As it uses the most basic implementation of search, I am not subdividing it further. Simple `Search` just runs this function from the root node